Library

rengarth opened the pull request:
Добавил базовый функционал, поигрался с HTML-шаблонами, после правок готов работать над конкретной реализацией библиотеки

rengarth committed:
Перезалив

Добавил базовый функционал, поигрался с HTML-шаблонами, добавил базовый функционал, после правок готов работать над конкретной реализацией библиотеки

rengarth committed:
Add files via upload

Не работает авторизация, приду - исправлю

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,42 @@
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
+        auth.jdbcAuthentication()
+                .dataSource(dataSource)
+                .passwordEncoder(NoOpPasswordEncoder.getInstance())
```
create bean and don`t use NoOp (use Bcrypt instead)

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,42 @@
+
+
+    @Override
+    protected void configure(AuthenticationManagerBuilder auth) throws Exception {
```
don`t do that - read about UserDetailsService

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+public class BookController {
+
+    @Autowired
+    private BookRepository bookRepository;
```
don`t user property autowiring - use setter or @RequiredArgsConstructor

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.*;
+
+@Controller
```
Must be RestController

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+    @Autowired
+    private BookRepository bookRepository;
+
+    @GetMapping("/add")
```
Create additional Mvc controller if you new mvc in your project

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @PostMapping("/add")
+    public String addBookPost(@RequestParam String name,
```
You`re using post - it gives you opportunity to use @RequestBody instead of multiple @RequestParam

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,35 @@
+@Controller
+public class RegistrationController {
+    @Autowired
+    private UserRepository userRepository;
```
don`t user property autowiring - use setter or @RequiredArgsConstructor

Kabbbie commented on src/main/java/com/libtask/library2/controllers/UserController.java:
```diff
@@ -0,0 +1,22 @@
+public class UserController {
+
+    @Autowired
+    UserRepository userRepository;
```
don`t user property autowiring - use setter or @RequiredArgsConstructor

Kabbbie commented on src/main/java/com/libtask/library2/entities/Book.java:
```diff
@@ -0,0 +1,51 @@
+        return getClass().hashCode();
+    }
+
+    public static String generateISBN() {
```
ISBN must not be generated - it`s standart so it need to be added into db manually

Kabbbie commented on src/main/java/com/libtask/library2/entities/Genre.java:
```diff
@@ -0,0 +1,15 @@
+    ROMANCE("Романтика"),
+    OTHER("Другое");
+
+    Genre(String genreName) {
```
access level

rengarth replied:
unnecessary (tip by IDEA)

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+    @NotNull Iterable<Book> findAll();
+    Optional<Book> findByName(String name);
+    Optional<Book> findByAuthorContains(String name);
+    @NonNull Optional<Book> findById(@NonNull String id);
```
this method already defined in CrudRepository

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+@Repository
+public interface BookRepository extends CrudRepository<Book, String> {
+
+    @NotNull Iterable<Book> findAll();
```
this method already defined in CrudRepository

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+import java.util.Optional;
+
+@Repository
+public interface BookRepository extends CrudRepository<Book, String> {
```
use JPARepository

Kabbbie commented on src/main/java/com/libtask/library2/repositories/UserRepository.java:
```diff
@@ -0,0 +1,17 @@
+@Repository
+public interface UserRepository extends CrudRepository<User, String> {
+
+    @NotNull Iterable<User> findAll();
```
this method already defined in CrudRepository

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
```
user RequiredArgsConstructor to prevent property injection problems in future

Kabbbie replied:
still don`t use @RequiredArgsConstructor

rengarth replied:
this is how the problem was solved in the controller, but this is already irrelevant, but the annotation remains

Kabbbie reviewed (changes requested):
Also add liquibase to easy data migration for your db

rengarth committed:
Add files via upload

Kabbbie commented on src/main/java/com/libtask/library2/Library2Application.java:
```diff
@@ -0,0 +1,13 @@
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+
+@SpringBootApplication
+public class Library2Application {
```
rename main class

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,70 @@
+        UserDetails user0 =
+                User.withDefaultPasswordEncoder()
+                        .username("amirlox@mail.ru")
+                        .password("1234")
```
didn`t find encoder bean

rengarth replied:
I will create it later when I move to storing logins and passwords in the database (soon!)

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping("/add")
+    public String addBook(Model model) {
+        return "add-book";
```
This is rest controller now. Why do we need api method that returns string

Kabbbie replied:
до сих пор не решено

rengarth replied:
я здесь ничего не сделаю без фронта в этом запросе

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+                              Model model) {
+        Book book = new Book(id, name, author, Genre.valueOf(genre), null);
+        bookRepository.save(book);
+        return "redirect:/books/catalog";
```
still works like mvc controller

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping("/catalog")
+    public String catalog(Model model) {
+        return String.valueOf(bookRepository.findAll());
```
why do you use String as boxed type? RestController will serialize object and returns it to browser as json

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+
+    @GetMapping("/{id}")
+    public String bookInfo(@PathVariable(value = "id") String id, Model model) {
+        return bookRepository.findById(id).toString();
```
why do you use String as boxed type? RestController will serialize object and returns it to browser as json

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @GetMapping("/{id}")
+    public String bookInfo(@PathVariable(value = "id") String id, Model model) {
```
model is redundant

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+
+
+    @GetMapping("/catalog")
+    public String catalog(Model model) {
```
model is redundant

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+                              String name,
+                              String author,
+                              String genre,
+                              Model model) {
```
model is redundant

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,43 @@
+    private BookRepository bookRepository;
+
+    @GetMapping("/add")
+    public String addBook(Model model) {
```
model is redundant

Kabbbie commented on src/main/java/com/libtask/library2/controllers/HomeController.java:
```diff
@@ -0,0 +1,14 @@
+
+    @GetMapping("/")
+    public String greeting(Model model) {
+        return "home";
```
This is rest controller now. Why do we need api method that returns string

Kabbbie replied:
also you already defined this redirect in MvcConfig

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,31 @@
+@RestController
+@RequiredArgsConstructor
+public class RegistrationController {
+    private UserRepository userRepository;
```
field might be final

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,31 @@
+        return "registration";
+    }
+
+    /*
```
don`t store commented code in repository

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,31 @@
+
+    @GetMapping("/registration")
+    public String registration() {
+        return "registration";
```
This is rest controller now. Why do we need api method that returns string

Kabbbie commented on src/main/java/com/libtask/library2/controllers/UserController.java:
```diff
@@ -0,0 +1,22 @@
+@RequestMapping("/users")
+public class UserController {
+    
+    UserRepository userRepository;
```
field might be private final

Kabbbie commented on src/main/java/com/libtask/library2/controllers/UserController.java:
```diff
@@ -0,0 +1,22 @@
+    UserRepository userRepository;
+
+    @GetMapping("/all")
+    public String showAllUsers (Model model) {
```
model is redundant

Kabbbie commented on src/main/java/com/libtask/library2/controllers/UserController.java:
```diff
@@ -0,0 +1,22 @@
+
+    @GetMapping("/all")
+    public String showAllUsers (Model model) {
+            return String.valueOf(userRepository.findAll());
```
why do you use String as boxed type? RestController will serialize object and returns it to browser as json

Kabbbie commented on src/main/java/com/libtask/library2/entities/User.java:
```diff
@@ -0,0 +1,38 @@
+    String email;
+
+    @Override
+    public boolean equals(Object o) {
```
use @EqualAndHashCode instead

rengarth replied:
IDE marks that with "Replace @EqualsAndHashCode with recommended equals() and hashCode() implementations." notification

Kabbbie commented on src/main/java/com/libtask/library2/entities/Book.java:
```diff
@@ -0,0 +1,44 @@
+    String userId;
+
+    @Override
+    public boolean equals(Object o) {
```
use @EqualAndHashCode instead

rengarth replied:
IDE marks that with "Replace @EqualsAndHashCode with recommended equals() and hashCode() implementations." notification

Kabbbie commented on src/main/java/com/libtask/library2/services/UserService.java:
```diff
@@ -0,0 +1,12 @@
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
```
user RequiredArgsConstructor to prevent property injection problems in future

rengarth committed:
Add files via upload

-Предыдущие замечания исправлены или прокомментированы
-Наведен структурный порядок в классах
-Добавлены native-query вызовы в интерфейсы соответствующих репозиториев

rengarth committed:
Delete Library2Application.java

rengarth committed:
Delete RegistrationController.java

rengarth committed:
Add files via upload

-Наведен порядок в структурах, исправлены предыдущие замечания
-Добавлен Encoder для in-memory юзеров
-Добавлен отдельный сервис и контроллер для учета книг

rengarth committed:
Add files via upload

rengarth committed:
Delete BalanceRepository.java

rengarth committed:
Reg & Auth

Add registration and auth by database

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,54 @@
+@RequiredArgsConstructor
+public class WebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @NonNull
```
Зачем? Он и так не будет пустым

rengarth replied:
Чтобы не возникало ошибок, связанных с аннотацией @RequiredArgsConstructor

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BalanceController.java:
```diff
@@ -0,0 +1,26 @@
+
+    private final BalanceService balanceService;
+
+    @PutMapping("/book/take")
```
1. Используй @RequestMapping чтобы не писать в каждом эндпоинте одинаковую часть
2. Почему класс называется BalanceController, а в эндпоинте слово book? Почитай про rest naming

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,38 @@
+
+@RestController
+@AllArgsConstructor
+@RequestMapping("/books")
```
book

rengarth replied:
rest naming говорит об ином

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,38 @@
+
+    @GetMapping("/add")
+    public void addBook() {
+        //необходима реализация фронта
```
Какая реализация фронта? Почему ты использовал GET метод для добавления?

rengarth committed:
Исправление ошибок

Ошибки исправлены, базовая работоспособность проверена. Тесты в ближайшее время.

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BalanceController.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @PutMapping("/book/return")
+    public User returnBook (Book book, User user) {
```
это не будет работать - где тесты?

Kabbbie replied:
Плюсом запрашиваемые данные избыточны - зачем мне объекты целиком, если мне хватит их id чтобы достать объекты из БД?

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BalanceController.java:
```diff
@@ -0,0 +1,26 @@
+    private final BalanceService balanceService;
+
+    @PutMapping("/book/take")
+    public User takeBook (Book book, User user) {
```
это не будет работать - где тесты?

Kabbbie replied:
Плюсом запрашиваемые данные избыточны - зачем мне обхекты целиком, если мне хватит их id чтобы достать объекты из БД?

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,25 @@
+
+    private final RegistrationService registrationService;
+
+    @GetMapping(path = "/reg")
```
не используй сокращения

Kabbbie replied:
зачем rest controller возвращает эту строку?

rengarth replied:
как заглушка

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,25 @@
+        return "let's sign up!";
+    }
+
+    @PostMapping(path = "/reg")
```
не используй сокращения

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,26 @@
+    private UserRepository userRepository;
+
+    public User takeBook(User user, Book bookOnBalance) {
+        userRepository.takeBook(user.getId(), bookOnBalance.getId());
```
Термин `takeBook` работает в контексте сервисного метода. В случае репозитория название метода должно описывать операцию, как пример addBookToUserBalance

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,26 @@
+        return user;
+    }
+
+    public User returnBook(User user, Book bookToReturn) {
```
Термин `returnBook` работает в контексте сервисного метода. В случае репозитория название метода должно описывать операцию, как пример deleteBookFromUserBalance

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,26 @@
+    private UserRepository userRepository;
+
+    public User takeBook(User user, Book bookOnBalance) {
+        userRepository.takeBook(user.getId(), bookOnBalance.getId());
```
Нет никакой валидации и проверок - существует ли книга/пользователь с такими данными
Без них я могу передать с фронта что угодно и будут созданы невалидные записи в бд

rengarth replied:
запрос на уровне БД не будет выполнен, записи не случится, случится ответ 500

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    public User returnBook(User user, Book bookToReturn) {
+        userRepository.returnBook(user.getId(), bookToReturn.getId());
```
Нет никакой валидации и проверок - существует ли книга/пользователь с такими данными, лежит ли книга в данный момент на балансе пользователя
Без них я могу передать с фронта что угодно и при попытке удаления всплывает конфликт - в лучшем случае увидим на фронте необработанную 500

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    @PostMapping(path = "/reg")
+    public User register(@RequestBody RegistrationRequest request) {
```
Зачем мы отдаем пользователя на фронт? Достаточно вернуть id, чтобы выполнить редирект в личный кабинет.

rengarth replied:
редирект будет выполнен уже на уровне фронта, я правильно понимаю?

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/entities/Book.java:
```diff
@@ -0,0 +1,38 @@
+    @Id
+    @Column(name = "id", unique = true, nullable = false)
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    Integer id;
```
Никогда не используй для числовых айдишников Integer как тип - используй Long

Kabbbie replied:
Почему не uuid?

rengarth replied:
Чтобы опробовать автоинкрементирование через БД и для того, чтобы наблюдать результат сверок было проще наглядно (что, конечно, необязательно, но это учебный проект)

Kabbbie commented on src/main/java/com/libtask/library2/entities/User.java:
```diff
@@ -0,0 +1,71 @@
+    @Id
+    @Column(name = "id", unique = true, nullable = false)
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    Integer id;
```
Никогда не используй для числовых айдишников Integer как тип - используй Long

Kabbbie replied:
Почему не uuid?

rengarth replied:
Чтобы опробовать автоинкрементирование через БД и для того, чтобы наблюдать результат сверок было проще наглядно (что, конечно, необязательно, но это учебный проект)

Kabbbie commented on src/main/java/com/libtask/library2/services/RegistrationService.java:
```diff
@@ -0,0 +1,26 @@
+    private final UserService userService;
+
+    public User register(RegistrationRequest request) {
+        User newUser = new User (
```
В текущей логике я могу зарегистрировать нескольких пользователей по одному email - думаю не стоить объяснять почему это плохо

rengarth replied:
я исправил всё согласно логике

Kabbbie commented on src/main/java/com/libtask/library2/services/RegistrationService.java:
```diff
@@ -0,0 +1,26 @@
+                request.getPassword(),
+                Role.USER
+        );
+        userService.signUpUser(newUser);
```
эта логика должна быть внутри RegistrationService

Kabbbie commented on src/main/java/com/libtask/library2/services/UserService.java:
```diff
@@ -0,0 +1,49 @@
+               throw new IllegalStateException("email already taken");
+          }
+
+          String encodedPassword = WebSecurityConfig.passwordEncoder().encode(user.getPassword());
```
Зачем ты создавал бин PasswordEncoder? Чтобы его вызывать через геттер? Используй DI

rengarth committed:
Add files via upload

Исправление ошибок, добавление функционала

rengarth committed:
Update BalanceController.java

rengarth committed:
Update BalanceController.java

rengarth committed:
Update BalanceController.java

rengarth committed:
Добавлены сервис-тесты

Добавлены базовые тесты сервисных классов

rengarth committed:
Delete HomeController.java

rengarth committed:
Delete BookEntity.java

rengarth committed:
Delete Genre.java

rengarth committed:
Delete UserEntity.java

rengarth committed:
Delete LibraryApplication.java

rengarth committed:
Delete BookRepository.java

rengarth committed:
Delete UserRepository.java

rengarth committed:
Delete LibraryApplication.java

rengarth committed:
Delete BookService.java

rengarth committed:
Delete UserService.java

rengarth committed:
Delete library_books.csv

rengarth committed:
Delete library_schema.drawio

rengarth committed:
Delete library_users.csv

rengarth committed:
Delete application.yml

rengarth committed:
Delete home.html

rengarth committed:
Delete LibraryApplicationTests.java

rengarth committed:
Add files via upload

Kabbbie commented on src/test/java/com/libtask/library2/services/BalanceServiceTest.java:
```diff
@@ -0,0 +1,88 @@
+                bookService.getBookByIsbn(book.getIsbn()));
+
+        //then
+        assertThat(bookService.getBookByIsbn(book.getIsbn()).getUserId()
```
можно использовать assertEquals()

Kabbbie commented on src/test/java/com/libtask/library2/services/BalanceServiceTest.java:
```diff
@@ -0,0 +1,88 @@
+                userService.getUserByEmail(user.getEmail()),
+                bookService.getBookByIsbn(book.getIsbn()));
+        //then
+        assertThat(bookService.getBookByIsbn(book.getIsbn()).getUserId() == null).isTrue();
```
можно использовать assertNull()

Kabbbie commented on src/test/java/com/libtask/library2/services/BalanceServiceTest.java:
```diff
@@ -0,0 +1,88 @@
+                userService.getUserByEmail(user.getEmail()),
+                bookService.getBookByIsbn(book.getIsbn()));
+
+        bookService.deleteBook(bookService.getBookByIsbn(book.getIsbn()));
```
почитай про @After для тестов

Kabbbie commented on src/test/java/com/libtask/library2/services/BalanceServiceTest.java:
```diff
@@ -0,0 +1,88 @@
+        //then
+        assertThat(bookService.getBookByIsbn(book.getIsbn()).getUserId() == null).isTrue();
+
+        bookService.deleteBook(bookService.getBookByIsbn(book.getIsbn()));
```
почитай про @After для тестов

Kabbbie commented on src/test/java/com/libtask/library2/services/BookServiceTest.java:
```diff
@@ -0,0 +1,50 @@
+
+        //then
+        assertThat(bookService.showCatalog().contains(bookService.getBookByIsbn(book.getIsbn()))).isTrue();
+        bookService.deleteBook(bookService.getBookByIsbn(book.getIsbn()));
```
почитай про @After для тестов

Kabbbie commented on src/main/java/com/libtask/library2/Main.java:
```diff
@@ -1,13 +1,13 @@
 
 @SpringBootApplication
-public class LibraryApplication {
+public class Main {
```
Да на самом деле название LibraryApp было бы норм

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/config/PasswordEncoder.java:
```diff
@@ -0,0 +1,14 @@
+public class PasswordEncoder {
+
+    @Bean
+    public static BCryptPasswordEncoder bCryptPasswordEncoder() {
```
method name might be just passwordEncoder

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/config/PasswordEncoder.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
+
+@Configuration
+public class PasswordEncoder {
```
Не обязательно выносить в отдельный конфигурационный файл. Можно оставить в конфиге безопасности

rengarth replied:
Исправлено, класс оставлен в конфиге безопасности. Именно класс, а не метод, для сохранения DI. Не удален на данный момент из проекта, так как остался незакрытый коммент

Kabbbie replied:
можно удалять сразу - коммент останется в summary к ревью и мне будет все видно

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,41 @@
+
+@RestController
+@AllArgsConstructor
+@RequestMapping("/books")
```
uri might be just "/book"

rengarth replied:
rest naming указывает, что именовать лучше следующим образом:
общий раздел, условно "/books", что-то из него конкретное, будь то каталог, сортировка или одна книга - уже через "/books/catalog", "books/{id}" и т.п.

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,24 @@
+
+    Book getBookByIsbn(String isbn);
+    @Query(value = "SELECT * FROM book", nativeQuery = true)
+    List<Book> showAllBooks();
```
JpaRepository уже имеет метод getAll()

rengarth replied:
findAll(), да. Исправлено.

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,24 @@
+    List<Book> showAllBooks();
+
+    @Query(value = "SELECT * FROM book WHERE user_id IS NULL", nativeQuery = true)
+    List<Book> getBooksInStock();
```
findFreeBooks

rengarth replied:
Исправлено, действительно, так звучит проще

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,24 @@
+    List<Book> getBooksInStock();
+
+    @Query(value = "SELECT * FROM book WHERE user_id = :user_id", nativeQuery = true)
+    List<Book> getTakenBooksListByUserId(@Param("user_id") Long userId);
```
findTakenBooksByUserId

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/repositories/UserRepository.java:
```diff
@@ -0,0 +1,41 @@
+@Repository
+public interface UserRepository extends JpaRepository<User, Long> {
+
+    User findByEmailIgnoreCase(String email);
```
а тут куда Optional делся?

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,37 @@
+@RequiredArgsConstructor
+public class BalanceService {
+
+    @NonNull
```
Подумай почему тут не нужна эта аннотация

rengarth replied:
Потому что я не создаю объект BalanceService balanceService явно, за меня это делает SpringBoot, мне же достаточно просто указать наличие поля данного класса в классе, где я буду функционал этого сервиса применять + потому что userRepo, как и bookRepo являются связанными с БД объектами, которые по умолчанию не могут быть null
@NonNull нужна для того, чтобы пометить аннотации @RequiredArgsConstructor, какие поля нужно указывать в конструкторе, который он создаст, а какие указать значением поля по умолчанию. Здесь была другая ошибка. Поля не были обозначены как final. Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,37 @@
+        if (!userRepository.existsById(user.getId())
+                || !bookRepository.existsById(bookToTake.getId())
+                || bookRepository.findById(bookToTake.getId()).get().getUserId() != null) {
+            throw new IllegalStateException("не соблюдены условия выполнения");
```
Exception message must be written in english

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,37 @@
+    private BookRepository bookRepository;
+
+    public void addBookToUserBalance(User user, Book bookToTake) {
+        if (!userRepository.existsById(user.getId())
```
Подумай над тем, почему так писать не надо и как этого избежать

rengarth replied:
проверка избыточна, можно вызвать объект Optional<T>, который в случае null или несовпадения и так выкинет ошибку. Тогда if сокращается до простого искомого условия: если поле user_id у книги != null.
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,105 @@
+        return newBook;
+    }
+
+    public void deleteBook(Book book) {
```
use dto instead of entity object

rengarth replied:
Так как у нас реализована логика множества копий одной книги в библиотеке, все её параметры (и даже isbn, если издания одинаковы) будут одинаковыми, в таком случае, при применении Dto возникнет закономерная ошибка логики исполнения. Вместо этого я сделал иначе - я передаю в метод Long id, уникальный номер каждой книги вообще, она подтягивается из БД и удаляется.

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,105 @@
+
+    public List<Book> sortByGenre(List<Book> bookList, Genre genre) {
+        return bookList.stream()
+                .filter(book -> book.getGenre().equals(genre))
```
это не сортировка, а фильтрация

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    //собственное решение
+    public List<List<Book>> pagination(List<Book> bookList, int pageValue) {
```
Что одно, что другое решение пагинации - плохое, так как пагинация должна происходить со стороны БД. В твоем случае на базу всегда дается полная нагрузка по получению полного списка

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,105 @@
+
+    public static List<Book> sortByAuthor(List<Book> bookList, String author) {
+        return bookList.stream()
+                .filter(book -> book.getAuthor().equalsIgnoreCase(author))
```
это не сортировка, а фильтрация

rengarth committed:
Add files via upload

На данный момент - без постраничного разбития и без изменения и добавления тестов. Это в процессе. Поправил и переделал по замечаниям и в целом.

rengarth committed:
Delete BookDto.java

rengarth committed:
Delete RegistrationRequest.java

rengarth committed:
Add files via upload

rengarth committed:
Delete Main.java

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,60 @@
+    public static class PasswordEncoder {
+
+        @Bean
+        public static BCryptPasswordEncoder passwordEncoder() {
```
бин не может быть статик - это не будет работать

rengarth replied:
Исправлено

Kabbbie commented on src/main/java/com/libtask/library2/config/WebSecurityConfig.java:
```diff
@@ -0,0 +1,60 @@
+    private final BCryptPasswordEncoder passwordEncoder;
+
+    @Configuration
+    public static class PasswordEncoder {
```
конфиг не может быть статик классом

rengarth replied:
Исправлено, класс объявлен вне класса WebSecurityConfig

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,57 @@
+        return bookService.getCatalog(page);
+    }
+
+    @GetMapping("/catalog/sorted-by-name")
```
Плохой вариант реализации различных сортировок. Поле по которому производится сортировка можно просто параметризовать в запросе и в итоге получить 1 эндпоинт вместо 3+

rengarth replied:
исправлено

Kabbbie commented on src/main/java/com/libtask/library2/controllers/BookController.java:
```diff
@@ -0,0 +1,57 @@
+        return bookService.bookToBookDto(bookService.getBookById(id));
+    }
+
+    @PutMapping("/delete")
```
Зачем прописывать delete в пути запроса, если есть DELETE метод в HTTP?
Хватило бы простого @DeleteMapping("/{id}")

Kabbbie commented on src/main/java/com/libtask/library2/controllers/HomeController.java:
```diff
@@ -0,0 +1,9 @@
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class HomeController {
```
Код не используется, но все еще лежит в реквесте - убрать

rengarth replied:
+

Kabbbie commented on src/main/java/com/libtask/library2/controllers/RegistrationController.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final RegistrationService registrationService;
+
+    @GetMapping(path = "/registration")
```
У обоих методов одинаковый путь - можно вынести в реквест маппинг

Kabbbie commented on src/main/java/com/libtask/library2/controllers/UserController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @GetMapping("/{id}/balance")
+    public Page<Book> showUserBalance(
```
Я бы перенес это в баланс контроллер

Kabbbie commented on src/main/java/com/libtask/library2/repositories/BookRepository.java:
```diff
@@ -0,0 +1,23 @@
+
+    Optional<Book> getBookByIsbn(String isbn);
+
+    @Query(value = "SELECT * FROM book WHERE user_id IS NULL ORDER BY id", nativeQuery = true)
```
ORDER BY id лишнее

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,34 @@
+    private final BookRepository bookRepository;
+
+    public void addBookToUserBalance(User user, Book bookToTake) {
+        User userInDb = userRepository.findById(user.getId()).orElseThrow();
```
fill orEsleThrow method

Kabbbie commented on src/main/java/com/libtask/library2/services/BalanceService.java:
```diff
@@ -0,0 +1,34 @@
+    private final UserRepository userRepository;
+    private final BookRepository bookRepository;
+
+    public void addBookToUserBalance(User user, Book bookToTake) {
```
User and Book as object is too much

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,63 @@
+                book.getGenre());
+    }
+
+    public Page<Book> getCatalog(Pageable page) {
```
лишнаяя функциональная обвязка вокруг репозитория

Kabbbie commented on src/main/java/com/libtask/library2/services/BookService.java:
```diff
@@ -0,0 +1,63 @@
+    public void deleteBook(Long id) {
+        Book book = getBookById(id);
+        if (book.getUserId() != null) {
+            throw new RuntimeException("Can't delete book that already taken");
```
specify an exception

Kabbbie commented on src/main/java/com/libtask/library2/services/RegistrationService.java:
```diff
@@ -0,0 +1,37 @@
+
+    public User register(RegistrationRequest request) {
+        User newUser = new User (
+                request.getFirstName(),
```
check incoming data

Kabbbie commented on src/main/java/com/libtask/library2/services/RegistrationService.java:
```diff
@@ -0,0 +1,37 @@
+        return newUser;
+    }
+
+    public void signUpUser(User user) {
```
private

Kabbbie commented on src/main/java/com/libtask/library2/services/UserService.java:
```diff
@@ -0,0 +1,48 @@
+     private final UserRepository userRepository;
+
+     public UserDto userToUserDto(User user) {
+          return new UserDto(
```
конструктор можно внести в класс

Kabbbie commented on src/main/resources/application.properties:
```diff
@@ -1 +1,11 @@
+spring.datasource.url=jdbc:postgresql://localhost:5432/Library
```
use application.yml instead of application.properties

Kabbbie commented on src/test/java/com/libtask/library2/services/BalanceServiceTest.java:
```diff
@@ -0,0 +1,84 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+
+@SpringBootTest
```
use dirty context

Kabbbie commented on src/test/java/com/libtask/library2/services/BookServiceTest.java:
```diff
@@ -0,0 +1,55 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+@SpringBootTest
+class BookServiceTest {
```
create additional tests

rengarth committed:
Delete PasswordEncoder.java

rengarth committed:
Delete HomeController.java

rengarth committed:
Доработка бэк-энд составляющей

unit-тестирование в скором времени

rengarth committed:
Add files via upload

rengarth committed:
unit-тесты для сервисов

rengarth committed:
Delete BookAddAndDeleteTest.java

rengarth committed:
Delete UserServiceTest.java

rengarth committed:
Delete registration.html

rengarth committed:
Delete greeting.html

rengarth committed:
Delete application.properties

rengarth committed:
Add files via upload

rengarth committed:
Add files via upload

rengarth committed:
Testing Git functions + minor changes

rengarth committed:
Merge remote-tracking branch 'origin/Library2' into Library2

# Conflicts:
#	src/main/java/com/libtask/library2/controllers/BalanceController.java
#	src/main/resources/static/js/userBalance.js

rengarth committed:
Testing Git functions + minor changes

rengarth committed:
1. Added Update Book Function (back-end);
2. Minor back-end changes;
3. Added sorting direction change by second click on direct head of column(front-end);
4. Minor UI features (front-end);

rengarth committed:
Delete .gitignore

rengarth committed:
Delete compiler.xml

rengarth committed:
Delete dataSources.xml

rengarth committed:
Delete encodings.xml

rengarth committed:
Delete jarRepositories.xml

rengarth committed:
Delete jpa-buddy.xml

rengarth committed:
Delete misc.xml

rengarth committed:
Delete vcs.xml

rengarth committed:
Delete MvcConfig.class

rengarth committed:
Delete WebSecurityConfig$PasswordEncoder.class

rengarth committed:
Delete WebSecurityConfig.class

rengarth committed:
Delete LibraryApp.class

rengarth committed:
Delete BalanceController.class

rengarth committed:
Delete BookController.class

rengarth committed:
Delete RegistrationController.class

rengarth committed:
Delete UserController.class

rengarth committed:
Delete WebController.class

rengarth committed:
Delete BookDto.class

rengarth committed:
Delete RegistrationRequest.class

rengarth committed:
Delete UserDto.class

rengarth committed:
Delete Book.class

rengarth committed:
Delete Genre.class

rengarth committed:
Delete GenreConverter.class

rengarth committed:
Delete Role.class

rengarth committed:
Delete User.class

rengarth committed:
Delete BalanceConditionException.class

rengarth committed:
Delete RegistrationComplianceException.class

rengarth committed:
Delete BookRepository.class

rengarth committed:
Delete UserRepository.class

rengarth committed:
Delete BalanceService.class

rengarth committed:
Delete BookService.class

rengarth committed:
Delete RegistrationService.class

rengarth committed:
Delete UserService.class

rengarth committed:
Delete application.yml

rengarth committed:
Delete add-book.html

rengarth committed:
Delete balance.html

rengarth committed:
Delete catalog-free.html

rengarth committed:
Delete catalog.html

rengarth committed:
Delete home.html

rengarth committed:
Delete login.html

rengarth committed:
Delete style.css

rengarth committed:
Delete books.js

rengarth committed:
Delete booksFree.js

rengarth committed:
Delete userBalance.js

rengarth committed:
1. Added condition safety while using other functions (frond-end);
2. Added tab "Пользователи" with basic functionality in it

rengarth committed:
1. Added condition safety while using other functions (frond-end);
2. Added tab "Пользователи" with basic functionality in it

rengarth committed:
Add files via upload

rengarth committed:
Test

rengarth committed:
Merge remote-tracking branch 'origin/Library2' into Library2

# Conflicts:
#	src/main/resources/static/js/users.js
#	src/main/resources/templates/users-list.html
#	target/classes/com/libtask/library2/config/MvcConfig.class
#	target/classes/com/libtask/library2/config/WebSecurityConfig$PasswordEncoder.class
#	target/classes/com/libtask/library2/config/WebSecurityConfig.class
#	target/classes/com/libtask/library2/controllers/WebController.class
#	target/classes/com/libtask/library2/repositories/BookRepository.class
#	target/classes/static/js/books.js
#	target/classes/static/js/booksFree.js
#	target/classes/static/js/userBalance.js
#	target/classes/templates/add-book.html
#	target/classes/templates/balance.html
#	target/classes/templates/catalog.html
#	target/classes/templates/home.html

rengarth committed:
Test2

rengarth committed:
Revert "Delete Book.class"

This reverts commit c4bf5f6a4714ef4910a09227287b843db2362424.

rengarth committed:
Revert "Delete UserDto.class"

This reverts commit 00123a6ff069fefa07646add0d369de339f3a51e.

rengarth committed:
Revert "Delete RegistrationRequest.class"

This reverts commit ac6d7d1b98a0c4eefcb6aca404398cf6234b983d.

rengarth committed:
Revert "Delete BookDto.class"

This reverts commit a522a2347d08d4542638b9b5a9b11598661a69ea.

rengarth committed:
Revert "Delete WebController.class"

This reverts commit 09f934bc

rengarth committed:
Revert "Delete .gitignore"

This reverts commit 19db096e10c2c788fc71c38ab3add53b0d49966b.

rengarth committed:
Revert "Delete jpa-buddy.xml"

This reverts commit 292bec5caea772039ecffd0d8b3c2c258c708700.

rengarth committed:
Revert "Delete BalanceController.class"

This reverts commit 6247f6e573f9e49f1513e897026fe9851e33fd14.

rengarth committed:
Revert "Delete BookController.class"

This reverts commit 6ebd94fa9a28a684027c0a4dbb8c44ce9140232e.

rengarth committed:
Revert "Delete RegistrationController.class"

This reverts commit 8e9cd92b466ad30dc43a441b18714c45768f2a3b.

rengarth committed:
Revert "Delete UserController.class"

This reverts commit 5810cd4e2cf34f7b462070dbc4cb938e99c4433f.

rengarth committed:
Revert "Delete Genre.class"

This reverts commit 62baf8fd44667ed4fc8d11ee0572d4ba609165fc.

rengarth committed:
Revert "Delete GenreConverter.class"

This reverts commit 9a7dbead737a4c57dbcccfa2538909c4b3819753.

rengarth committed:
Revert "Delete Role.class"

This reverts commit 9fd72ac928f8b11a390e3cd1ed2fa2eeafc167b4.

rengarth committed:
Revert "Delete User.class"

This reverts commit 7bcfdf55bb9079a9878be1a22a1dbbb3cd068eee.

rengarth committed:
Revert "Delete BalanceConditionException.class"

This reverts commit 14d7543509bb2b1d992f8b061111931aa495909a.

rengarth committed:
Revert "Delete RegistrationComplianceException.class"

This reverts commit 771b5812fe0f193e1df783f810b7450cc0701544.

rengarth committed:
Revert "Delete BookService.class"

This reverts commit 9e841d02cb31309bf88239f6e6cffe8522230eea.

rengarth committed:
Revert "Delete RegistrationService.class"

This reverts commit fbd5855c6418bda9a765a8154352c14a432d4714.

rengarth committed:
Revert "Delete UserService.class"

This reverts commit b90008852e43e84b15c48e2495999daf25277610.

rengarth committed:
Revert "Delete catalog-free.html"

This reverts commit a9ab1935d95d4dca51745f85cb0181f5114500e1.

rengarth committed:
Add files via upload

rengarth committed:
Revert "Revert "Delete WebController.class""

This reverts commit ab213127dab38ae2402c10948fbf8c93f7b71a1d.

rengarth committed:
Delete MvcConfig.class

rengarth committed:
Delete WebSecurityConfig$PasswordEncoder.class

rengarth committed:
Delete WebSecurityConfig.class

rengarth committed:
Delete WebController.class

rengarth committed:
Delete BookRepository.class

rengarth committed:
Delete books.js

rengarth committed:
Delete booksFree.js

rengarth committed:
Delete userBalance.js

rengarth committed:
Delete users.js

rengarth committed:
Delete add-book.html

rengarth committed:
Delete balance.html

rengarth committed:
Delete catalog.html

rengarth committed:
Delete home.html

rengarth committed:
Delete users-list.html

rengarth committed:
Full functional on Books

rengarth committed:
Changes

rengarth committed:
Add .gitignore

rengarth committed:
Delete .gitignore

rengarth committed:
Delete jpa-buddy.xml

rengarth committed:
Delete MvcConfig.class

rengarth committed:
Delete BalanceController.class

rengarth committed:
Delete BookController.class

rengarth committed:
Delete RegistrationController.class

rengarth committed:
Delete UserController.class

rengarth committed:
Delete BookDto.class

rengarth committed:
Delete RegistrationRequest.class

rengarth committed:
Delete UserDto.class

rengarth committed:
Delete Book.class

rengarth committed:
Delete Genre.class

rengarth committed:
Delete GenreConverter.class

rengarth committed:
Delete Role.class

rengarth committed:
Delete User.class

rengarth committed:
Delete BalanceConditionException.class

rengarth committed:
Delete RegistrationComplianceException.class

rengarth committed:
Delete BookService.class

rengarth committed:
Delete RegistrationService.class

rengarth committed:
Delete UserService.class

rengarth committed:
Delete books.js

rengarth committed:
Delete booksFree.js

rengarth committed:
Delete userBalance.js

rengarth committed:
Delete balance.html

rengarth committed:
Delete catalog-free.html

rengarth committed:
Delete catalog.html

rengarth committed:
Add docker-compose

rengarth committed:
Merge remote-tracking branch 'origin/Library2' into Library2

# Conflicts:
#	.idea/jpa-buddy.xml

rengarth committed:
Add docker-compose enhanced

rengarth committed:
Work in progress

rengarth committed:
Final Commit

rengarth committed:
Final Commit 2
